### ISSUE TITLE
Grafana chart is not upgrade-able due to pvc storageclass conflict

### ISSUE THREAD

**Issue by dcherniv** [helm/charts#20389]
Output of `helm version`:
```
version.BuildInfo{Version:"v3.0.2", GitCommit:"19e47ee3283ae98139d98460de796c1be1e3975f", GitTreeState:"clean", GoVersion:"go1.13.5"}
```
Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```
Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
EKS, GKE.

To reproduce. Install grafana from official repo. Enable persistence
```
dchernivetsky@XQ8SVXR8X339:~$ helm3 install grafana stable/grafana --set persistence.enabled=true
NAME: grafana
LAST DEPLOYED: Mon Jan 27 12:00:40 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
NOTES:
1. Get your 'admin' user password by running:

   kubectl get secret --namespace default grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo

2. The Grafana server can be accessed via port 80 on the following DNS name from within your cluster:

   grafana.default.svc.cluster.local

   Get the Grafana URL to visit by running these commands in the same shell:

     export POD_NAME=$(kubectl get pods --namespace default -l "app=grafana,release=grafana" -o jsonpath="{.items[0].metadata.name}")
     kubectl --namespace default port-forward $POD_NAME 3000

3. Login with the password from step 1 and the username: admin
dchernivetsky@XQ8SVXR8X339:~$ 
```
Confirm that persistent volume claim got created with standard class:
```
dchernivetsky@XQ8SVXR8X339:~$ kubectl get pvc 
NAME      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
grafana   Bound    pvc-a2ff28cb-4126-11ea-97ad-42010a8c400b   10Gi       RWO            standard       52s
dchernivetsky@XQ8SVXR8X339:~$ 
```
Upgrade the chart:
```
dchernivetsky@XQ8SVXR8X339:~$ helm3 upgrade grafana stable/grafana --set persistence.enabled=true
Error: UPGRADE FAILED: cannot patch "grafana" with kind PersistentVolumeClaim: PersistentVolumeClaim "grafana" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound 
claims
dchernivetsky@XQ8SVXR8X339:~$ 
```
This issue is in relation to https://github.com/helm/helm/issues/6642

**Comment by bacongobbler** [comment]
transferring to helm/charts as this appears to be in relation to a specific chart's usage.

**Comment by dcherniv** [comment]
@bacongobbler This is not related to specific chart usage. This is a general problem with the way helm 3 does 3-way diff.
Please see https://github.com/helm/helm/issues/6642#issuecomment-542839246

**Comment by acim** [comment]
Is this a helm3 issue or not, this is easy fixable and should be fixed in the chart like other charts handle this. If you look at pvc.yaml of Prometheus, there are conditions:

`{{- if .Values.server.persistentVolume.storageClass }}
{{- if (eq "-" .Values.server.persistentVolume.storageClass) }}
  storageClassName: ""
{{- else }}
  storageClassName: "{{ .Values.server.persistentVolume.storageClass }}"
{{- end }}
{{- end }}`

**Comment by kenperkins** [comment]
This appears to have a proposed fix in https://github.com/helm/charts/pull/17806

**Comment by dcherniv** [comment]
@acim 
That conditional in prometheus wouldn't work because the class is still set to "" which depending on the cloud and admission controller may or may not work. 
On GKE with storageClass as ""
```
NAME      STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
grafana   Pending                                                     3s
```
```  
Normal  FailedBinding  11s (x3 over 26s)  persistentvolume-controller  no persistent volumes available for this claim and no storage class is set

```
@kenperkins #17806 looks better and should work properly, though there may be some side effects which are not entirely clear to me, based on the below:
https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1

@bacongobbler 
On a side note, i get that helm 3 is a major upgrade, but requiring all users of your software to change their behavior for something that is quite trivial to fix and something that was never against the kubernetes spec in the first place, which helm is expected to honor, is just not how you develop.
It has the effect of turning users away, who may start looking for other solutions, like kustomize.

**Comment by acim** [comment]
@dcherniv, this means that GKE doesn't set default Storageclass. Well, I am not aware of that, but AKS and EKS work fine with the code like in Prometheus. 

https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/

**Comment by acim** [comment]
Hmm, GKE documentation says that they actually define the default StorageClass.

```Volume implementations such as gcePersistentDisk are configured through StorageClass resources. GKE creates a default StorageClass for you which uses the standard persistent disk type (ext4). The default StorageClass is used when a PersistentVolumeClaim doesn't specify a StorageClassName.```

https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes

**Comment by dcherniv** [comment]
@acim the code in Prometheus does specify storageclass as "", which is different than omitting storageClass field altogether, which appears to be different than setting storageClass as 'storageClass: '

**Comment by acim** [comment]
I agree with that and didn't say you should copy that code, but there should be some conditionals for ommiting the StorageClass completely. However, Prometheus chart works well on EKS without defining the StorageClass value, meaning this "" works. I saw more charts using "".

**Comment by dcherniv** [comment]
@acim @bacongobbler
I further tested this a little bit. It appears that 
```
'storageClass: '
```
is really not a valid spec or shouldn't be a valid spec really. The proper way to go about selecting default storage class is omitting storageClass field altogether.
The approach taken by prometheus chart is incorrect. Specifying storageClass as "" will cause kubernetes (in some cloud provider) to specifically look for storageClass "" which may or may not exist.

TL;DR fix in https://github.com/helm/charts/pull/17806 is correct.

**Comment by eloo** [comment]
Is this really fixed?
If i create a new deployment with the helm chart and want to update it without any changes i still get this error.

**Comment by iptizer** [comment]
> Is this really fixed?
> If i create a new deployment with the helm chart and want to update it without any changes i still get this error.

I don't see it changed: [https://github.com/helm/charts/blob/master/stable/grafana/templates/pvc.yaml#L26](https://github.com/helm/charts/blob/master/stable/grafana/templates/pvc.yaml#L26)

And I don't see a way to easily work around this when it is set for once, other that explicitly setting it to the value.

**Comment by NeckBeardPrince** [comment]
@iptizer Getting the same thing on the latest version
